### ISSUE TITLE
The provided new optimizer is sensitive on tiny batchsize

### ISSUE THREAD

**Issue by GreatGBL** [Luolc/AdaBound#11]
The provided new optimizer is sensitive on tiny batchsize (<4), I am testing on the very simply linear regression, while others performance looks like nice currently.

Path:

![1](https://user-images.githubusercontent.com/10858450/53678905-48d64c00-3cbd-11e9-9583-4cf56c8af760.png)

Loss curve:
![2](https://user-images.githubusercontent.com/10858450/53678907-4d026980-3cbd-11e9-97bb-e889923b8c89.png)

Zoomed Loss curve:
![3](https://user-images.githubusercontent.com/10858450/53678910-4ffd5a00-3cbd-11e9-88f3-15fc1ed8b182.png)

**Comment by Luolc** [comment]
That's very interesting. We didn't pay attention to the impact by batch size before.
Thanks for providing the new aspect to explore! 😄 

Would you please provide more details of the experiments? Such as the hyperparameters of each optimizer, the scale of the dataset, etc.

**Comment by GreatGBL** [comment]
#!/usr/bin/env python
# coding: utf-8

# In[104]:


import torch
import numpy as np
import matplotlib.pyplot as plt
import torch.nn.functional as F
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader


## generate M data points roughly forming a line (noise added)
M = 50
theta_true = torch.Tensor([[0.5], [2]])

X = 10 * torch.rand(M, 2) - 5
X[:, 1] = 1.0

y = torch.mm(X, theta_true) + 0.3 * torch.randn(M, 1)


# In[105]:


def mse(t1, t2):
    diff = t1 - t2
    return torch.sum(diff * diff) / diff.numel()

def model(x):
    return X @ theta

def cost_func(theta, X, y): 
    # YOUR CODE HERE
#     raise NotImplementedError()
        
    pred =  torch.mm(X, theta)
    diff =  pred - y
    loss =  (diff**2).sum(0) / X.shape[0]
    return loss


# In[106]:


# Define
batch_size = 16
num_epochs = 100

loss_fn = F.mse_loss

train_ds = TensorDataset(X, y)
train_dl = DataLoader(train_ds, batch_size, shuffle=True)
model = nn.Linear(2, 1, bias=False)


# In[107]:


import math
import torch
from torch.optim import Optimizer


class AdaBound(Optimizer):
    """Implements AdaBound algorithm.
    It has been proposed in `Adaptive Gradient Methods with Dynamic Bound of Learning Rate`_.
    Arguments:
        params (iterable): iterable of parameters to optimize or dicts defining
            parameter groups
        lr (float, optional): Adam learning rate (default: 1e-3)
        betas (Tuple[float, float], optional): coefficients used for computing
            running averages of gradient and its square (default: (0.9, 0.999))
        final_lr (float, optional): final (SGD) learning rate (default: 0.1)
        gamma (float, optional): convergence speed of the bound functions (default: 1e-3)
        eps (float, optional): term added to the denominator to improve
            numerical stability (default: 1e-8)
        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
        amsbound (boolean, optional): whether to use the AMSBound variant of this algorithm
    .. Adaptive Gradient Methods with Dynamic Bound of Learning Rate:
        https://openreview.net/forum?id=Bkg3g2R9FX
    """

    def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), final_lr=0.1, gamma=1e-3,
                 eps=1e-8, weight_decay=0, amsbound=False):
        if not 0.0 <= lr:
            raise ValueError("Invalid learning rate: {}".format(lr))
        if not 0.0 <= eps:
            raise ValueError("Invalid epsilon value: {}".format(eps))
        if not 0.0 <= betas[0] < 1.0:
            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
        if not 0.0 <= betas[1] < 1.0:
            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))
        if not 0.0 <= final_lr:
            raise ValueError("Invalid final learning rate: {}".format(final_lr))
        if not 0.0 <= gamma < 1.0:
            raise ValueError("Invalid gamma parameter: {}".format(gamma))
        defaults = dict(lr=lr, betas=betas, final_lr=final_lr, gamma=gamma, eps=eps,
                        weight_decay=weight_decay, amsbound=amsbound)
        super(AdaBound, self).__init__(params, defaults)

        self.base_lrs = list(map(lambda group: group['lr'], self.param_groups))

    def __setstate__(self, state):
        super(AdaBound, self).__setstate__(state)
        for group in self.param_groups:
            group.setdefault('amsbound', False)

    def step(self, closure=None):
        """Performs a single optimization step.
        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            loss = closure()

        for group, base_lr in zip(self.param_groups, self.base_lrs):
            for p in group['params']:
                if p.grad is None:
                    continue
                grad = p.grad.data
                if grad.is_sparse:
                    raise RuntimeError(
                        'Adam does not support sparse gradients, please consider SparseAdam instead')
                amsbound = group['amsbound']

                state = self.state[p]

                # State initialization
                if len(state) == 0:
                    state['step'] = 0
                    # Exponential moving average of gradient values
                    state['exp_avg'] = torch.zeros_like(p.data)
                    # Exponential moving average of squared gradient values
                    state['exp_avg_sq'] = torch.zeros_like(p.data)
                    if amsbound:
                        # Maintains max of all exp. moving avg. of sq. grad. values
                        state['max_exp_avg_sq'] = torch.zeros_like(p.data)

                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
                if amsbound:
                    max_exp_avg_sq = state['max_exp_avg_sq']
                beta1, beta2 = group['betas']

                state['step'] += 1

                if group['weight_decay'] != 0:
                    grad = grad.add(group['weight_decay'], p.data)

                # Decay the first and second moment running average coefficient
                exp_avg.mul_(beta1).add_(1 - beta1, grad)
                exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
                if amsbound:
                    # Maintains the maximum of all 2nd moment running avg. till now
                    torch.max(max_exp_avg_sq, exp_avg_sq, out=max_exp_avg_sq)
                    # Use the max. for normalizing running avg. of gradient
                    denom = max_exp_avg_sq.sqrt().add_(group['eps'])
                else:
                    denom = exp_avg_sq.sqrt().add_(group['eps'])

                bias_correction1 = 1 - beta1 ** state['step']
                bias_correction2 = 1 - beta2 ** state['step']
                step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1

                # Applies bounds on actual learning rate
                # lr_scheduler cannot affect final_lr, this is a workaround to apply lr decay
                final_lr = group['final_lr'] * group['lr'] / base_lr
                lower_bound = final_lr * (1 - 1 / (group['gamma'] * state['step'] + 1))
                upper_bound = final_lr * (1 + 1 / (group['gamma'] * state['step']))
                step_size = torch.full_like(denom, step_size)
                step_size.div_(denom).clamp_(lower_bound, upper_bound).mul_(exp_avg)

                p.data.add_(-step_size)

        return loss


# In[108]:


# Define a utility function to train the model
def fit(num_epochs, loss_fn, opt):
    model.weight.data[0][0].fill_(2.00)
    model.weight.data[0][1].fill_(4.00)
    Loss =[]
    Theta = np.zeros(shape=(1,2,num_epochs)) 
    for epoch in range(num_epochs):
        for xb,yb in train_dl:
            # Generate predictions
            pred = model(xb)
            loss = loss_fn(pred, yb)
            # Perform gradient descent
            loss.backward()
            opt.step()
            opt.zero_grad()
    
        Loss.append(loss_fn(model(X), y))
        
        Theta[:,0,epoch] = model.weight.detach().numpy()[0][0]
        Theta[:,1,epoch] = model.weight.detach().numpy()[0][1]
        
    Loss = np.array(Loss)
    
    return Theta, Loss


# In[109]:


ADAM_t, ADAM = fit(num_epochs, loss_fn, torch.optim.Adam(model.parameters(), lr=1e-2) )
SGD_t, SGD = fit(num_epochs, loss_fn, torch.optim.SGD(model.parameters(), lr=1e-3, momentum=0) )
SGDM_t, SGDM = fit(num_epochs, loss_fn, torch.optim.SGD(model.parameters(), lr=1e-3, momentum=0.9) )
ADAB_t, ADAB = fit(num_epochs, loss_fn, AdaBound(model.parameters(), lr=1e-2, final_lr=0.1) )


# In[110]:


theta_0_vals = np.linspace(-2,4,100)
theta_1_vals = np.linspace(0,4,100)

theta = torch.Tensor(len(theta_0_vals),2)
J = np.zeros((len(theta_0_vals),len(theta_1_vals)))
# The following for loops are used to compute the value of the cost function, J, over all the thetas in order to plot the contour below.
for i, theta_0 in enumerate(theta_0_vals):
    for j, theta_1 in enumerate(theta_1_vals):
        ## YOUR CODE HERE
        J[i,j] = cost_func(torch.Tensor([[theta_0], [theta_1]]), X, y)

xc,yc = np.meshgrid(theta_0_vals, theta_1_vals)
contours = plt.contour(xc, yc, J, 20)

plot_vals = range(0,num_epochs)
plt.plot(ADAM_t[0,0,plot_vals],ADAM_t[0,1,plot_vals],'-.',lw=2, label='Adam')
plt.plot(SGD_t[0,0,plot_vals],SGD_t[0,1,plot_vals],'-.',lw=2, label='Sgd')
plt.plot(SGDM_t[0,0,plot_vals],SGDM_t[0,1,plot_vals],'-.',lw=2, label='Sgd+momentum')
plt.plot(ADAB_t[0,0,plot_vals],ADAB_t[0,1,plot_vals],'-.',lw=2, label='AdaBound')

plt.scatter(theta_true[0].numpy(),theta_true[1].numpy(),marker='*', color='red',lw=2, label='gloal')

plt.legend(loc='lower left')


# In[111]:


plt.figure()
plt.subplot(211)
plt.plot(range(ADAB.shape[0]),ADAB,'-.',lw=2, label='AdaBound')
plt.plot(range(ADAB.shape[0]),ADAM,'-.',lw=2, label='Adam')
plt.plot(range(ADAB.shape[0]),SGD,'-.',lw=2, label='Sgd')
plt.plot(range(ADAB.shape[0]),SGDM,'-.',lw=2, label='Sgd+momentum')
plt.subplots_adjust(top=2.92, bottom=0.12, left=0.15, right=2.95, hspace=0.2, wspace=0.35)
plt.legend(loc='upper right')

plt.subplot(212)
plt.plot(range(ADAB.shape[0]),ADAB,'-.',lw=2, label='AdaBound')
plt.plot(range(ADAB.shape[0]),ADAM,'-.',lw=2, label='Adam')
plt.plot(range(ADAB.shape[0]),SGD,'-.',lw=2, label='Sgd')
plt.plot(range(ADAB.shape[0]),SGDM,'-.',lw=2, label='Sgd+momentum')
plt.subplots_adjust(top=2.92, bottom=0.12, left=0.15, right=2.95, hspace=0.2, wspace=0.35)
plt.xlim((80, 100))
plt.ylim((0, 0.3))
plt.legend(loc='upper right')


# In[112]:


print('AdaBound',ADAB[ADAB.shape[0]-2:ADAB.shape[0]])
print('--')
print('ADAM',ADAM[ADAB.shape[0]-2:ADAB.shape[0]])
print('--')
print('SGD',SGD[ADAB.shape[0]-2:ADAB.shape[0]])
print('--')
print('sgd+momentum',SGDM[ADAB.shape[0]-2:ADAB.shape[0]])


# In[ ]:

**Comment by GreatGBL** [comment]
import torch
import numpy as np
import matplotlib.pyplot as plt
import torch.nn.functional as F
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader
import math
import torch
from torch.optim import Optimizer

## generate M data points roughly forming a line (noise added)

M = 50

theta_true = torch.Tensor([[0.5], [2]])

X = 10 * torch.rand(M, 2) - 5

X[:, 1] = 1.0

y = torch.mm(X, theta_true) + 0.3 * torch.randn(M, 1)


def mse(t1, t2):

    diff = t1 - t2
	
    return torch.sum(diff * diff) / diff.numel()

def model(x):

    return X @ theta

def cost_func(theta, X, y):  

    pred =  torch.mm(X, theta)
	
    diff =  pred - y
	
    loss =  (diff**2).sum(0) / X.shape[0]
	
    return loss



# Define

batch_size = 1

num_epochs = 100

loss_fn = F.mse_loss

train_ds = TensorDataset(X, y)

train_dl = DataLoader(train_ds, batch_size, shuffle=True)

model = nn.Linear(2, 1, bias=False)

# Define a utility function to train the model
def fit(num_epochs, loss_fn, opt):

    model.weight.data[0][0].fill_(2.00)
	
    model.weight.data[0][1].fill_(4.00)
	
    Loss =[]
	
    Theta = np.zeros(shape=(1,2,num_epochs)) 
	
    for epoch in range(num_epochs):
	
        for xb,yb in train_dl:
		
            pred = model(xb)
			
            loss = loss_fn(pred, yb)
			
            loss.backward()
			
            opt.step()
			
            opt.zero_grad()
    
        Loss.append(loss_fn(model(X), y))
        
        Theta[:,0,epoch] = model.weight.detach().numpy()[0][0]
		
        Theta[:,1,epoch] = model.weight.detach().numpy()[0][1]
        
    Loss = np.array(Loss)
    
    return Theta, Loss

ADAM_t, ADAM = fit(num_epochs, loss_fn, torch.optim.Adam(model.parameters(), lr=1e-2) )

SGD_t, SGD = fit(num_epochs, loss_fn, torch.optim.SGD(model.parameters(), lr=1e-3, momentum=0) )
SGDM_t, SGDM = fit(num_epochs, loss_fn, torch.optim.SGD(model.parameters(), lr=1e-3, momentum=0.9) )

ADAB_t, ADAB = fit(num_epochs, loss_fn, AdaBound(model.parameters(), lr=1e-2, final_lr=0.1) )


theta_0_vals = np.linspace(-2,4,100)

theta_1_vals = np.linspace(0,4,100)

theta = torch.Tensor(len(theta_0_vals),2)
J = np.zeros((len(theta_0_vals),len(theta_1_vals)))

for i, theta_0 in enumerate(theta_0_vals):

    for j, theta_1 in enumerate(theta_1_vals):
	

        J[i,j] = cost_func(torch.Tensor([[theta_0], [theta_1]]), X, y)

xc,yc = np.meshgrid(theta_0_vals, theta_1_vals)

contours = plt.contour(xc, yc, J, 20)

plot_vals = range(0,num_epochs)

plt.plot(ADAM_t[0,0,plot_vals],ADAM_t[0,1,plot_vals],'-.',lw=2, label='Adam')

plt.plot(SGD_t[0,0,plot_vals],SGD_t[0,1,plot_vals],'-.',lw=2, label='Sgd')

plt.plot(SGDM_t[0,0,plot_vals],SGDM_t[0,1,plot_vals],'-.',lw=2, label='Sgd+momentum')

plt.plot(ADAB_t[0,0,plot_vals],ADAB_t[0,1,plot_vals],'-.',lw=2, label='AdaBound')

plt.scatter(theta_true[0].numpy(),theta_true[1].numpy(),marker='*', color='red',lw=2, label='gloal')

plt.legend(loc='lower left')


plt.figure()

plt.subplot(211)

plt.plot(range(ADAB.shape[0]),ADAB,'-.',lw=2, label='AdaBound')

plt.plot(range(ADAB.shape[0]),ADAM,'-.',lw=2, label='Adam')

plt.plot(range(ADAB.shape[0]),SGD,'-.',lw=2, label='Sgd')

plt.plot(range(ADAB.shape[0]),SGDM,'-.',lw=2, label='Sgd+momentum')

plt.subplots_adjust(top=2.92, bottom=0.12, left=0.15, right=2.95, hspace=0.2, wspace=0.35)

plt.legend(loc='upper right')

plt.subplot(212)

plt.plot(range(ADAB.shape[0]),ADAB,'-.',lw=2, label='AdaBound')

plt.plot(range(ADAB.shape[0]),ADAM,'-.',lw=2, label='Adam')

plt.plot(range(ADAB.shape[0]),SGD,'-.',lw=2, label='Sgd')

plt.plot(range(ADAB.shape[0]),SGDM,'-.',lw=2, label='Sgd+momentum')

plt.subplots_adjust(top=2.92, bottom=0.12, left=0.15, right=2.95, hspace=0.2, wspace=0.35)

plt.xlim((80, 100))

plt.ylim((0, 0.3))

plt.legend(loc='upper right')

**Comment by stevenyangyj** [comment]
When I used AdaBound to train a ShuffleNet V2 model with tiny batch (5-10), I met same problem. This optimizer might not be convergent.

Btw: when I used "adabound.AdaBound([{'params': part of model's params, 'lr':0...}])" to prevent some parameters be updated during training process, I would get error infomation means "cannot use lr = 0". But I can use "torch.optim.Adam([{'params': part of model's params, 'lr':0...}])" to implement this purpose.

Is this a BUG ??

**Comment by felipevw** [comment]
Hi,
I have been training with adabound in a custom dataset and I faced similar issues with low batch sizes.
The only doubt I have is that in the ReadMe, you provide a comparison graph of the different optimizers, I dont understand why the abrupt change in the epoch 150. I guess there is when the optimizers switches to SGD but why at this point? Does that mean that if I train a dataset through 1000 epochs, it will make a similar change in epoch 750?

Thank you for the help
